### PR TITLE
[graphql] Add graphql filter with multiple values

### PR DESCRIPTION
**Pull request by ArnoudThibaut**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | no
| Fixed tickets | -
| License       | MIT
| Doc PR        | https://github.com/api-platform/docs/pull/524

Using the SearchFilter with multiple values was not possible with graphql.

Using the same synthax than the query filter (`foo[]=bar`) was not possible because characters `[` and `]` in a graphql query are invalid and it seems that it's not possible to use the same key for both single value filter and multi values filter.

The idea is to prefix the field name with a special keyword to make the distinction between single and multiple values.

### CODE REVIEW

**Comment by dunglas** [comment]
Good one, it can even be considered a bug fix to me. Regarding the name, it will be cleaner and safer to add a separator, I would also use the term `list` instead of array, as in the GraphQL spec.

So I would rename `arrayrelatedDummy_name` to `relatedDummy_name_list`. WDYT?

**Comment by ArnoudThibaut** [comment]
Yep, `list` is better than array.

For the separator, `_` is already used by nested properties, could it be a problem to use it for that too?

**Comment by dunglas** [comment]
I don't think so. At worst, for a property called `list`, it would generate something like `class_list_list`. It will still work and it cannot have conflicts.

**Comment by ArnoudThibaut** [comment]
Ok, the change is done.

**Comment by ArnoudThibaut** [comment]
Yes sure,
I will make a PR for the doc tonight or tomorrow.

File: src/GraphQl/Type/SchemaBuilder.php [hunk 1]
```diff
@@ -231,12 +231,15 @@ private function getResourceFieldConfiguration(string $resourceClass, ResourceMe
                         $filterType = \in_array($value['type'], Type::$builtinTypes, true) ? new Type($value['type'], $nullable) : new Type('object', $nullable, $value['type']);
                         $graphqlFilterType = $this->convertType($filterType, false, null, $depth);
 
-                        if ('[]' === $newKey = substr($key, -2)) {
-                            $key = $newKey;
+                        if ('[]' === substr($key, -2)) {
                             $graphqlFilterType = GraphQLType::listOf($graphqlFilterType);
+                            $key = substr($key, 0, -2).'_list';
```
Comment (toofff) [inline src/GraphQl/Type/SchemaBuilder.php]: `$key = "{${substr($key, 0, -2)}}_list";`?

Comment (dunglas) [reply src/GraphQl/Type/SchemaBuilder.php]: It's not in JS :D

Comment (toofff) [reply src/GraphQl/Type/SchemaBuilder.php]: Hooo shit :smile:

**Comment by ArnoudThibaut** [comment]
Doc PR just here : https://github.com/api-platform/docs/pull/524

File: src/GraphQl/Type/SchemaBuilder.php [hunk 2]
```diff
@@ -231,12 +231,15 @@ private function getResourceFieldConfiguration(string $resourceClass, ResourceMe
...
                         }
 
                         parse_str($key, $parsed);
+                        if (array_key_exists($key, $parsed) && \is_array($parsed[$key])) {
```
Comment (meyerbaptiste) [inline src/GraphQl/Type/SchemaBuilder.php]: `isset($parsed[$key])` will stop the condition earlier with a null value.

**Comment by dunglas** [comment]
Thanks @ArnoudThibaut

**Pull request merged**